### DISCUSSION TITLE
Examples for visualizing continuous space?

### DISCUSSION THREAD

**Discussion by PestyVesty** [mesa/mesa#2356] [Q&A] [answered]
Is there a simple example code for visualizing continuous space? So far I've tried different versions of mesa to run the boid flocker example. When using mesa 2.04 I would get the following error:
```
 File "C:\Users\x\Documents\Repos\x\boid_flockers\boid_flockers\model.py", line 50, in __init__
TypeError: Agent.__init__() missing 1 required positional argument: 'model'
```

And when I try to use mesa 3.0 I would get the following error:
```
 File "C:\Users\x\Documents\Repos\boid_flockers\boid_flockers\SimpleContinuousModule.py", line 4, in <module>
    class SimpleCanvas(mesa.visualization.VisualizationElement):
                       ^^^^^^^^^^^^^^^^^^
AttributeError: module 'mesa' has no attribute 'visualization'
```

Is there something that I'm missing? Is there an example code to visualize continuous space environments?

Thanks

**Comment by EwoutH** [comment, marked as answer]
Thanks for reaching out! The examples on the main mesa-examples branch are already updated for Mesa 3.0 (see https://github.com/projectmesa/mesa-examples/issues/148). You can either:
- Update to the latest Mesa 3.0 beta with `pip install --upgrade --pre mesa`. See the Mesa 3.0 [Migration guide](https://mesa.readthedocs.io/latest/migration_guide.html), [Release notes](https://github.com/projectmesa/mesa/blob/main/HISTORY.md) and [Mesa overview](https://mesa.readthedocs.io/latest/overview.html)
- The use an example that works with Mesa 2.x, from the [`mesa-2.x` branch](https://github.com/projectmesa/mesa-examples/issues/148) in mesa-examples.

Let me know if this helps!

**Reply by rht** [reply]
Even with `pip install --upgrade --pre mesa`, it won't be solved, because the example code still uses Tornado-based visualization.

**Reply by EwoutH** [reply]
BoidFlockers does have a SolaraViz implementation: [examples/boid_flockers/app.py](https://github.com/projectmesa/mesa-examples/blob/main/examples/boid_flockers/app.py)

**Reply by PestyVesty** [reply]
Thanks for the support, currently I got the visualisation for boid flockers working using Mesa 2.4.

I wasn't able to get the visualisation working for Mesa 3.0, I still get the same error as shown above. I assume the code for boid flockers is not been up to date for Mesa 3.0? Or I'm missing something?

(Regardless thanks for the help, I realized I didn't branch into mesa-2.x this entire time)

**Reply by EwoutH** [reply]
Glad to hear it works on 2.4!

I currently have an PR open that updates the example to both work with 3.0 and have improved visualization: https://github.com/projectmesa/mesa/pull/2365

**Reply by PestyVesty** [reply]
Hectic! Will definetly look into that first when I get the time. Thanks a lot.

**Reply by PestyVesty** [reply]
I got the visualization working for boid flockers using Mesa 3.0. Hope the following instructions can help others in the future:

1. Installed the mesa version on github using ```pip install -U -e git+https://github.com/projectmesa/mesa@main#egg=mesa```
2. Cloned the mesa repo from ```https://github.com/projectmesa/mesa```
3. In the mesa repo, navigated to ```examples/basic/boid_flocker```
4. There was no ```run.py```file, so I created a jupyter notebook file in ```boid_flocker```
5. Copy the code in ```app.py``` into a cell and run the cell

Question: is it possible to run the visualisation on a web browser instead of jupyter notebook?

**Reply by EwoutH** [reply]
Thanks for your steps!

> Question: is it possible to run the visualisation on a web browser instead of jupyter notebook?

Certainly! `cd examples/basic/boid_flockers` and then run `solara run app.py` in your terminal.

As you can see, this is all very active development, and we will also update the documentation ASAP.

https://github.com/projectmesa/mesa/pull/2381 is also interesting to follow.
